ui/README.md

```diff
@@ -1,8 +1,44 @@
 # VSCode Azure SDK for Node.js - UI Tools (Preview)
 
-This package provides common Azure UI elements for VS Code extensions.
+This package provides common Azure UI elements for VS Code extensions:
+* [AzureActionHandler](#azure-action-handler): Displays error messages and optionally adds telemetry to commands/events.
+* [AzureTreeDataProvider](#azure-tree-data-provider): Displays an Azure Explorer with Azure Subscriptions and child nodes of your implementation.
+* [AzureBaseEditor](#azure-base-editor): Displays a text editor with upload support to Azure.
 
-> NOTE: This package throws a `UserCancelledError` if the user cancels an operation. This error should be handled appropriately by your extension.
+> NOTE: This package throws a `UserCancelledError` if the user cancels an operation. If you do not use the AzureActionHandler, you must handle this exception in your extension.
+
+## Azure Action Handler
+
+Use the Azure Action Handler to consistently display error messages and track commands with telemetry. You should contruct the handler and register commands/events in your extension's `activate()` method. The simplest example is to register a command (in this case, refreshing a node):
+```typescript
+const actionHandler: AzureActionHandler = new AzureActionHandler(context, outputChannel, reporter);
+actionHandler.registerCommand('yourExtension.Refresh', (node: IAzureNode) => { node.refresh(); });
+```
+Here are a few of the benefits this provides:
+* Parses Azure errors of the form `{ "Code": "Conflict", "Message": "This is the actual message" }` and only displays the 'Message' property
+* Displays single line errors normally and multi-line errors in the output window
+* If you pass a TelemetryReporter, tracks multiple properties in addition to the [common extension properties](https://github.com/Microsoft/vscode-extension-telemetry#common-properties):
+  * result (Succeeded, Failed, or Canceled)
+  * duration
+  * error
+
+If you want to add custom telemetry proprties, use `registerCommandWithCustomTelemetry` and add your own properties or measurements:
+```typescript
+actionHandler.registerCommandWithCustomTelemetry('yourExtension.Refresh', async (properties: TelemetryProperties, measurements: TelemetryMeasurements) => {
+    properties.customProp = "example prop";
+    measurements.customMeas = 49;
+});
+```
+
+Finally, you can also register events. The main difference is that your callback must take in the `trackTelemetry` parameter. Events are not tracked by default (since they can happen very frequently). You must call `trackTelemetry()` to signal that this event is indeed handled by your extension. For example, if your extension only handles `json` files in the `onDidSaveTextDocument`, it might look like this:
```

Comment:
trackTelemetry sounds like a Boolean.  How about sendEventTelemetry or sendTelemetry?

Reply:
I'm worried 'sendTelemetry' implies it's actually sending the telemetry at that point in the code, so users will call it _after_ their own custom logic like this:
```
if (doc.fileExtension === 'json') {
    throw new Error('Oops I messed up');
    sendTelemetry();
}
```
In the above example, that means we don't catch their error in our telemetry because 'sendTelemetry' was never reached. If they did this, we would've tracked the error:
```
if (doc.fileExtension === 'json') {
    sendTelemetry();
    throw new Error('Oops I messed up');
}
```

I was kind of going for a boolean feeling with 'trackTelemetry' since it's basically flipping a switch saying "track telemetry for this event", but it's not actually _sending_ the telemetry until the command finishes